### ISSUE TITLE
aur-build: fails with "file exists" if makepkg --sign is used

### ISSUE THREAD

**Issue by je-vv** [AladW/aurutils#410]
Before current pacman upgrade to 5.1.0-1, I could make makepkg to sign packages by default, enabling "sign" to the BUILDENV variable of /etc/makepkg.conf, and still get away with NOT using "aursync -s ...", because I didn't want the database to be signed (any manual repo-add or repo-remove corrupts signing).  I got warning about the packages signatures, for some reason tried to be included in the repos, as if they were packages, which I honestly don't understand why, but it worked.

1st thing, is to avoid trying that, it's sort of weird, only the packages should be attempted to be added into the repos (signatures can be part of the package metadata I'd guess).

2nd, with the upgrade I'm not able to get away with signed packages anymore, I'm getting now errors rather than warnings, which is really sad:

+++
==> ERROR: 'package-query-1.9-3-x86_64.pkg.tar.xz.sig' is not a package file, skipping
+++

That used to be a warning, :(  Essentially I'm looking to deal correctly with signed packages, without the need to also sign the repo database.

I hope it's possible...  For now, it seems I need to stop signing packages...

Thanks !

**Comment by AladW** [comment]
I have no idea why you'd want to do this in the first place.

**Comment by maximbaz** [comment]
> because I didn't want the database to be signed (any manual repo-add or repo-remove corrupts signing)

A wild guess, maybe you are looking for the `-s` flag, i.e. `repo-add -s ...`? Otherwise I'm also not sure why you want this, I'd suggest investigating why your repo gets corrupted and fixing that instead.

**Comment by AladW** [comment]
Right, I seem to have missed this in the OP:

>any manual repo-add or repo-remove corrupts signing

In that case you should use `repo-add -sv` as @maximbaz suggested. This is the same command `aur-build` uses internally.

**Comment by AladW** [comment]
I guess if for whatever reason one used `makepkg --sign` - which is somewhat valid with `aur-build` taking arbitrary makepkg arguments after `--` - it will fail on repo-add due to `.sig` files ending in `PKGDEST`. I guess one could use `extglob` in that case:
```
shopt -s extglob
pkglist=(./*.!(sig))
```
Thoughts @maximbaz ?

**Comment by AladW** [comment]
Alternatively it could check for .sig in PKGDEST and bail if any are found.

**Comment by maximbaz** [comment]
Mmm, sure - I guess if passing `--sign` is allowed, we shouldn't bail on finding .sig files, but ignoring them should be fine.

**Comment by je-vv** [comment]
OK, thanks to all, sorry for not being able to answer any sooner...

So I guess it'll remain closed, and I learned how to manually add packages (with or without signature) to a signed repo DB I guess, :+1: 

I think I'll keep makepkg.conf without any hint to sign packages, and use aursync, aurbuild and others with "-s" then...  Though I really don't think I'll take advantage of having a repo DB signed (perhaps I'm missing something, and I might learn about it later)...

Thanks !

**Comment by AladW** [comment]
I'll probably revisit this topic if I decide to implement `generate_signature.sh` from `libmakepkg` which for some reason lacks `--batch` (resulting in gpg nagging if there's an existing signature) or have some different motivation to change the `aur-build` logic.

>Though I really don't think I'll take advantage of having a repo DB signed

http://www2.cs.arizona.edu/stork/packagemanagersecurity/attacks-on-package-managers.html#explanation

**Comment by je-vv** [comment]
BTW, I just tried then using "ursync -s -u --repo <my_repo>", but having as well in makepkg.conf:

 <BUILDENV=(!distcc color !ccache check sign)>

And I got:

"gpg: signing failed: File exists"
"gpg: signing failed: File exists"

So definitely if makepkg is configured to sign packages, still aurutils fail to work when signing packages as well.  My current work around it, is to create an alias for makepkg, so when running manually I don't forget to add "-s"...

Is this also something you would be looking at?

Thanks !

**Comment by AladW** [comment]
That sucks I guess.

**Comment by AladW** [comment]
@je-vv https://github.com/AladW/aurutils/commit/9905843722a2c3a4e3adac0b0f71c25f677c4435

Note that if signing is disabled in `aur-build`, existing signatures from BUILDENV or makepkg --sign are simply skipped.

**Comment by AladW** [comment]
Documented with: https://github.com/AladW/aurutils/commit/eee53d7af67eac14ec1e92ee5ce93ddfb4181453

**Comment by AladW** [comment]
Then again having such a policy in place complicates things so I'll leave it to users to shoot themselves in the foot. 0a253ef8f11588a25c65f239e76f7541abd434fb

**Comment by je-vv** [comment]
As it ended up, [https://github.com/AladW/aurutils/blob/master/lib/aur-build#L193-L203](url), is just fine.  If it finds the signatures, it uses them, if not, but aurutils were instructed to sign, then the packages will get signed...  That great...

The #https://github.com/AladW/aurutils/commit/eee53d7af67eac14ec1e92ee5ce93ddfb4181453 documentation sounds weird to me, though I'm not a native English speaker.  I would rephrase it like:

> If signing packages through "makepkg" whether using its "--sign" option or
> by its configuration variable "BUILDENV" ("makepkg.conf"), no new signature
> will be generated by aur-build, no matter it was called with "-s" or not, and those
> signatures will always be provided by aur-build along with the packages themselves.
> If no signatures are generated by "makepkg", and the "-s" option is given to
> aur-build, then those signatures packages are generated and provided by
> aur-build along with its corresponding packages.

Not sure if it's clearer or not.  As stated, I'm not a native English speaker, :).  So it's up to you to reconsider the documentation.  The code looks just great to me...

Thanks again !

**Comment by AladW** [comment]
>One additional question, what release target would you have for these changes?

I put it on the 2.0.0 milestone, which I'd say will take at least until next year to complete, if ever.

**Comment by maximbaz** [comment]
I thought I read somewhere that you wanted to align 2.0.0 with pacman 5.1 release 😜

**Comment by AladW** [comment]
Yeah, but there's so much stuff left to do that I don't think that's realistic...
